pvanalytics/features/shading.py

```diff
@@ -0,0 +1,338 @@
+
+    Returns
+    -------
+    int
+        The height of the bounding box in pixels.
+    """
+    return bbox[2] - bbox[0] + 1
+
+
+def _filter_blobs(wires, blob_length, connectivity):
+    """Remove blobs shorter than blob_length.
+
+    Examines all connected components in the image and removes any
+    that span fewer than `blob_length` columns.
+    """
+    components = measure.label(wires, connectivity=connectivity, background=0)
+    keep = [props.label for props in measure.regionprops(components)
+            if _height(props.bbox) > blob_length]
+    return np.isin(components, keep)
+
+
+def _tand(degrees):
+    """Tangent of an angle in degrees."""
+    return np.tan(np.deg2rad(degrees))
+
+
+def _filter_bars(wires, out):
+    """Only keep shadows that are 'bar-shaped' and span multiple days."""
+    for angle in range(0, 90, 5):
+        if angle < 75:
+            dim = 20
+            mid = 10
+        else:
+            dim = round(10 + 2 * angle // 5)
+            mid = round(6 + angle // 5)
+        cc = np.arange(0, dim)
+        se = np.zeros((dim, dim))
+        if angle <= 45:
+            rr = mid + np.round((cc - mid) * _tand(angle))
+            se[cc, rr.astype('int')] = 1
+        else:
+            rr = mid + np.round((cc - mid) * _tand(90 - angle))
+            se[cc, rr.astype('int')] = 1
+            se = se.T
+        out = np.logical_or(
+            out,
+            morphology.binary_opening(wires, se)
+        )
+        out = np.logical_or(
+            out,
+            morphology.binary_opening(wires, np.flipud(se))
+        )
+    return out
+
+
+def _clean_wires(wires):
+    """Clean up clouds that are connected to wires."""
+    out = _remove_pillars(wires)
+    out = _fill_gaps(out)
+    out = _remove_spikes(out)
+    out = _fill_gaps(out)
+    out = _restore_gaps(out)
+    out = _filter_blobs(out, 20, connectivity=1)
+    out = _filter_bars(wires, out)
+    out = _filter_blobs(out, 15, connectivity=2)
+    return out
+
+
+def fixed(ghi, daytime, clearsky, interval=None, min_gradient=2):
+    """Detects shadows form fixed structures such as wires and poles.
+
+    Uses morphological image processing methods to identify shadows
+    from fixed local objects in GHI data. GHI data are are assumed to
+    be reasonably complete with relatively few missing values and at a
+    fixed time interval nominally of 1 minute over the course of
+    several months. Shadows are assumed to be relatively short
+    duration. The algorithm forms a 2D image of the GHI data by
+    arranging time of day along the x-axis and day of year along the
+    y-axis. Rapid change in GHI in the x-direction is used to identify
+    edges of shadows; continuity in the y-direction is used to
+    separate local object shading from cloud shadows.
+
+    Parameters
+    ----------
+    ghi : Series
+        Time series of GHI measurements. Should be at 1-minute frequency
```

Comment:
I think we should make 1-minute frequency a requirement. I don't have any evidence that this algorithm works for slightly irregular time spacing, and the conversion to an image assumes that the data have a regular frequency.

Reply:
Makes my life easier. Should we require any particular number of days, or just sternly warn the users should have several months of data for this to work well?

Reply:
There's a vertical dimension to the elements used in `_filter_bars`. Maybe the code will run for data with fewer days, but that's not the intent of the algorithm. I'd warn that the algorithm should be applied to at least 60 days of data. If there's a hard lower limit on the vertical dimension of the image, from the `ndimage` functions, we could state that also. May not be worth figuring that limit out.